<a href="https://colab.research.google.com/github/wcoleman194754/CS6647/blob/main/CS_6647_Hw_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np

In [2]:
state = {
    0: 'rest',
    1: 'sweeties',
    2: 'exercises'
}
state

{0: 'rest', 1: 'sweeties', 2: 'exercises'}

In [3]:
# transitional matrix

M = np.array([[0.2,0.2,0.6], [0.2,0.1,0.7], [0.1,0.3,0.6]])
M

array([[0.2, 0.2, 0.6],
       [0.2, 0.1, 0.7],
       [0.1, 0.3, 0.6]])

In [4]:
# random walk

n=15
starting_state = 0
previous_state = starting_state

for i in range(n):
  current_state = np.random.choice([0,1,2],p=M[previous_state])
  print(state[current_state], "---->", end="")
  previous_state = current_state



exercises ---->sweeties ---->exercises ---->sweeties ---->exercises ---->sweeties ---->exercises ---->exercises ---->exercises ---->exercises ---->sweeties ---->exercises ---->exercises ---->rest ---->exercises ---->

In [14]:
# monte carlo method to determine probabilities

steps = 15
start_state = 0
pi = np.array([0,0,0])
pi[start_state] = 1
previous_state = start_state

for i in range(steps):
  current_state = np.random.choice([0,1,2], p=M[previous_state])
  pi[current_state]+=1
  previous_state = current_state

print("pi = ", pi/steps)

pi =  [0.2        0.33333333 0.53333333]


In [6]:
# monte carlo method to determine probabilities

steps = 10000
start_state = 0
pi = np.array([0,0,0])
pi[start_state] = 1
previous_state = start_state

for i in range(steps):
  current_state = np.random.choice([0,1,2], p=M[previous_state])
  pi[current_state]+=1
  previous_state = current_state

print("pi = ", pi/steps)

pi =  [0.1329 0.2384 0.6288]


In [10]:
# repeated matrix multiplication

steps = 15
M_n = M


for i in range(steps):
  M_n = np.matmul(M_n, M)

print("M^n = \n", M_n, "\n")
print("pi = ", M_n[0])


M^n = 
 [[0.13761468 0.23853211 0.62385321]
 [0.13761468 0.23853211 0.62385321]
 [0.13761468 0.23853211 0.62385321]] 

pi =  [0.13761468 0.23853211 0.62385321]


In [9]:
# repeated matrix multiplication

steps = 10000
M_n = M


for i in range(steps):
  M_n = np.matmul(M_n, M)

print("M^n = \n", M_n, "\n")
print("pi = ", M_n[0])

M^n = 
 [[0.13761468 0.23853211 0.62385321]
 [0.13761468 0.23853211 0.62385321]
 [0.13761468 0.23853211 0.62385321]] 

pi =  [0.13761468 0.23853211 0.62385321]


##observations

We can see the results of the state probabilities from matrix.
As the number of steps get large then the probability states converge on their values. If the simulations are low then the model tends to overstate the probabilities for rest and sweeties. As the number of simulations gets large then we can see that they settle on around 13.76% for rest, 23.85% for sweeties and 62.39% for exercises. 